stdlib/public/core/Integers.swift.gyb

```diff
@@ -3610,21 +3610,27 @@ extension ${Self} : Hashable {
   public var hashValue: Int {
     @inline(__always)
     get {
-% if bits <= word_bits and signed:
-      // Sign extend the value.
-      return Int(self)
-% elif bits <= word_bits and not signed:
-      // Sign extend the value.
-      return Int(${OtherSelf}(bitPattern: self))
-% elif bits == word_bits * 2:
-      // We have twice as many bits as we need to return.
-      return
-        Int(truncatingIfNeeded: self) ^
-        Int(truncatingIfNeeded: self &>> 32)
-% else:
-      _Unimplemented()
-% end
+      return _hashValue(for: self)
+    }
+  }
+
+  @_inlineable // FIXME(sil-serialize-all)
+  public func _hash(into hasher: _UnsafeHasher) -> _UnsafeHasher {
+    // FIXME(hasher): To correctly bridge `Set`s/`Dictionary`s containing
+    // `AnyHashable`-boxed integers, all integer values are currently required
+    // to hash exactly the same way as the corresponding (U)Int64 value.  To fix
+    // this, we should introduce a custom AnyHashable box for integer values
+    // that sign-extends values to 64 bits.
```

Comment:
I really don't think it's problematic that equal integer values hash the same way regardless of their bit width.

In fact, it helps us out nicely if ever we want to extend heterogeneous `==` beyond `BinaryInteger` types to, say, `Array<T> where T: BinaryInteger`. I absolutely agree that enabling per-execution variation in hash values for integers is strictly an improvement, but I don't think treating this as a fixme falls into the same category.

Reply:
It causes potential performance issues, because we're doing a full 64-bit SipHash `compress` round for every `UInt8` value, when we could batch eight of them up into a single round instead. This also applies to `Bool`s, small enums and the like. This effect needs to be measured to quantify it, though; it may not actually turn out to be significant. 

Another potential issue is that extending everything to 32/64 bits washes away distinctions between distinct types, perhaps slightly weakening SipHash.

An alternative is to analyze wide integer values and hash them at the shortest width that can contain them -- like the code currently does for `Int64` on 32-bit systems. I suspect this would have a measurable effect on Int.hashValue's performance, though -- but this needs to be verified. We'd also have to devise separate API for code that wants to feed a specific series of bytes to the hasher, with no such arbitrary preprocessing.